firedrake/mesh.py

```diff
@@ -2036,13 +2036,16 @@ def spatial_index(self):
         coords_min = self._order_data_by_cell_index(column_list, coords_min.dat.data_ro_with_halos)
         coords_max = self._order_data_by_cell_index(column_list, coords_max.dat.data_ro_with_halos)
 
-        # Push max and min out so we can find points on the boundary within
-        # tolerance. Note that if tolerance is too small it might not actually
-        # change the value!
-        if hasattr(self, "tolerance") and self.tolerance is not None:
-            coords_diff = coords_max - coords_min
-            coords_min -= self.tolerance*coords_diff
-            coords_max += self.tolerance*coords_diff
+        # Push max and min out to a hypercube centered at its average so we can
+        # find points on the mesh within tolerance.
```

Comment:
I think this could do with more explanation in the comment - having sat down with pen and paper I don't entirely understand what's going on. It relies on the pseudo-inverse you spoke about am I correct?

Reply:
No, it's just really basic geometry. The pseudo-inverse stuff only comes into play at the point where we find local coordinates (and was already working so isn't in this PR).